src/libstd/io/cursor.rs

```diff
@@ -0,0 +1,430 @@
+    pub fn into_inner(self) -> T { self.inner }
+
+    /// Get a reference to the underlying value in this cursor.
+    pub fn get_ref(&self) -> &T { &self.inner }
+
+    /// Get a mutable reference to the underlying value in this cursor.
+    ///
+    /// Care should be taken to avoid modifying the internal I/O state of the
+    /// underlying value as it may corrupt this cursor's position.
+    pub fn get_mut(&mut self) -> &mut T { &mut self.inner }
+
+    /// Returns the current value of this cursor
+    pub fn position(&self) -> u64 { self.pos }
+
+    /// Sets the value of this cursor
+    pub fn set_position(&mut self, pos: u64) { self.pos = pos; }
+}
+
+macro_rules! seek {
+    () => {
+        fn seek(&mut self, style: SeekFrom) -> io::Result<u64> {
+            let pos = match style {
+                SeekFrom::Start(n) => n as i64,
+                SeekFrom::End(n) => self.inner.len() as i64 + n,
+                SeekFrom::Current(n) => self.pos as i64 + n,
+            };
+
+            if pos < 0 {
+                Err(Error::new(ErrorKind::InvalidInput,
+                               "invalid seek to a negative position",
+                               None))
+            } else {
+                self.pos = pos as u64;
+                Ok(self.pos)
+            }
+        }
+    }
+}
+
+impl<'a> io::Seek for Cursor<&'a [u8]> { seek!(); }
+impl<'a> io::Seek for Cursor<&'a mut [u8]> { seek!(); }
+impl io::Seek for Cursor<Vec<u8>> { seek!(); }
+
+macro_rules! read {
+    () => {
+        fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+            if self.pos > self.inner.len() as u64 { return Ok(0) }
+            let mut slice = &self.inner[(self.pos as usize)..];
+            let n = try!(slice.read(buf));
+            self.pos += n as u64;
+            Ok(n)
+        }
+    }
+}
+
+impl<'a> Read for Cursor<&'a [u8]> { read!(); }
+impl<'a> Read for Cursor<&'a mut [u8]> { read!(); }
+impl Read for Cursor<Vec<u8>> { read!(); }
+
+macro_rules! buffer {
+    () => {
+        fn fill_buf(&mut self) -> io::Result<&[u8]> {
+            if self.pos < (self.inner.len() as u64) {
+                Ok(&self.inner[(self.pos as usize)..])
+            } else {
+                Ok(&[])
+            }
+        }
+        fn consume(&mut self, amt: usize) { self.pos += amt as u64; }
+    }
+}
+
+impl<'a> BufRead for Cursor<&'a [u8]> { buffer!(); }
+impl<'a> BufRead for Cursor<&'a mut [u8]> { buffer!(); }
+impl<'a> BufRead for Cursor<Vec<u8>> { buffer!(); }
+
+impl<'a> Write for Cursor<&'a mut [u8]> {
+    fn write(&mut self, data: &[u8]) -> io::Result<usize> {
+        if self.pos >= self.inner.len() as u64 { return Ok(0) }
+
+        let amt = {
+            let mut s = &mut self.inner[(self.pos as usize)..];
+            try!(s.write(data))
+        };
+        self.pos += amt as u64;
+        Ok(amt)
+    }
+}
+impl Write for Cursor<Vec<u8>> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
```

Comment:
It would be more helpful to me at least for this to be a little more constructive. Do you have some suggestions about how the method could be trimmed down?